README.md
# [quran-promise](https://github.com/axis-tip/quran-promise)

[![NPM version](http://img.shields.io/npm/v/quran-promise.svg?style=flat-square)](https://www.npmjs.com/package/quran-promise)
[![NPM downloads](http://img.shields.io/npm/dm/quran-promise.svg?style=flat-square)](https://www.npmjs.com/package/quran-promise)
[![Build Status](http://img.shields.io/travis/axis-tip/quran-promise/master.svg?style=flat-square)](https://travis-ci.org/axis-tip/quran-promise)
[![Coverage Status](https://img.shields.io/coveralls/axis-tip/quran-promise.svg?style=flat-square)](https://coveralls.io/github/axis-tip/quran-promise)
[![Dependency Status](http://img.shields.io/david/axis-tip/quran-promise.svg?style=flat-square)](https://david-dm.org/axis-tip/quran-promise)

> ES2016 async/await (promise) enabled node module for the Holy Quran. Uses the Tanzil corpus: http://tanzil.net.

## Using this Module

For detailed documentation, and code samples for how to use this module in your app or service please visit: https://axis-tip.github.io/quran-promise/.

## Developing this Module

The rest of this `README` describes how you can clone this repo to get the source to develop/test locally. Contributions are very welcome!

### Environment Setup

First, set up your dev tools and node.js:

1. Set up your favourite IDE. We use VSCode: https://code.visualstudio.com/ and you can configure this IDE as follows:
  1. Follow this useful article about JavaScript development in VSCode [here](https://code.visualstudio.com/docs/languages/javascript), including how to debug.
  2. Set up linting in VSCode. You can install eslint globally via `npm install -g eslint` and then install the VSCode `eslint` extension. See [here](http://stackoverflow.com/questions/36327096/vscode-linter-es6-es7-babel-linter) for more info.
2. Install Node.js (version 6.x or higher)
3. Install Git: https://git-scm.com/download

Next, clone the repo and install dependencies.

```sh
$ git clone https://github.com/axis-tip/quran-promise
$ cd quran-promise
$ npm install
```

The quran data is in a sqllite database, and a known good version is checked in. Optionally, ff you want to re-generate it, please run:

```sh
npm run gendb
```

Finally, if you're new to Next Generation JavaScript you might want to brush up on the following resources:

1. ES6 Training Course](https://es6.io/friend/konstantin) by Wes Bos
2. [You Don't Know JS: ES6 & Beyond](http://amzn.to/2bzvV51) by Kyle Simpson (Dec, 2015)
3. [Understand promises before you start using async/await](https://medium.com/@bluepnume/learn-about-promises-before-you-start-using-async-await-eb148164a9c8#.q2u0j68wd)

### Testing, Running and Publishing

Run one, or a combination of the following commands to lint and test your code:

```sh
$ npm run lint          # Lint the source code with ESLint (runs as part of CI build)
$ npm run lint:fix      # Lint the source code with ESLint, using the --fix option to auto-fix some issues
$ npm test              # Run unit tests with Mocha
$ npm run test:watch    # Run unit tests with Mocha, and watch files for changes
$ npm run test:cover    # Run unit tests with code coverage by Istanbul (runs as part of CI build)
$ npm run coveralls     # Report code coverage to coveralls.io (runs as part of CI build)
```

To launch the documentation site, run:

```sh
$ npm install -g easystatic
$ npm run start:docs    # Launches the docs site locally for testing
$ npm run publish:docs  # Optionally, publishes the doc site to GitHub pages (runs as part of CI build)
```

Finally, to build and publish to npm, you need to run:

```sh
npm run build           # Builds the source, transpiling via babel (runs as part of CI build)
cd dist                 # Publish from the dist folder
npm publish             # Requires credentials allowed to publish to npm
```

## Credits

With thanks, this module acknowledges and depends on the following resources:

1. Tanzil project: http://tanzil.net/
2. EveryAyah: http://everyayah.com/data/status.php
3. The Babel Starter Kit, used to scaffold the initial version of this module: https://github.com/kriasoft/babel-starter-kit/

Finally, a HUGE callout to the  similar npm module https://www.npmjs.com/package/quran. We ended up completely rewriting it,
to support ES2016 async/await (promises), but got inspired by it. We also tried to keep the calling semantics as compatible as possible
to aid anyone who might be migrating from that module.

### License

MIT © 2016 Axis, the Information Professionals.


CHANGELOG.md
## quran-promise Change Log

All notable changes to this project will be documented in this file.

### [v0.0.7] - 2016-09-15

- Fixed database relative path issues and tested in a separate node application
- Very basic documentation on the docs site, so people can start using this module

### [v0.0.3] - 2016-09-13

- Added basic query and search capabilities
- Automated tests at 100% coverage

### [v0.0.1] - 2016-09-09

- Initial scaffolding and repo structure (not a ton of functionality)


test/QuranSpec.js
/**
 * quran-promise
 *
 * Copyright © 2016 Axis, the Information Professionals. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import chai from 'chai';
import Quran from '../src';

// set up chai
const expect = chai.expect;
chai.use(require('chai-string'));

describe('Quran', () => {

  describe('Quran.get(...)', () => {

    it('should return a numbered verse of a numbered ayat', async () => {
      // Fetch verse 1 of Al Fatiha via get
      const quran = new Quran();
      const verses = await quran.get(1, 3);

      // TEST: Surah Fatiha has 7 ayahs, of which we are fetching only the 3rd
      expect(verses.length).to.be.equal(1);

      // TEST: Arabic text of the 3rd ayah of Surah Fatiha matches
      expect(verses[0]).to.be.equal('ٱلرَّحْمَٰنِ ٱلرَّحِيمِ');
    });

    it('should return a full numbered ayat', async () => {
      // Fetch Al Fatiha via get
      const quran = new Quran();
      const verses = await quran.get(1);

      // TEST: Surah Fatiha has 7 ayahs
      expect(verses.length).to.be.equal(7);

      // TEST: Arabic text of the 3rd ayah of Surah Fatiha matches
      expect(verses[2]).to.be.equal('ٱلرَّحْمَٰنِ ٱلرَّحِيمِ');
    });
  });

  describe('Quran.chapters(...)', () => {

    it('should return info about a chapter', async () => {
      // Fetch info about chapter 1
      const quran = new Quran();
      const chapters = await quran.chapters(1);

      // TEST: We are getting metadata for 1 chapter
      expect(chapters.length).to.be.equal(1);

      // TEST: Surah Fatiha is Meccan
      expect(chapters[0].type).to.be.equal('Meccan');
    });

    it('should return info about all chapters', async () => {
      // Fetch info about all chapters
      const quran = new Quran();
      const chapters = await quran.chapters();

      // TEST: There are 114 surahs
      expect(chapters.length).to.be.equal(114);

      // TEST: Surah Al-Nas is Meccan
      expect(chapters[chapters.length - 1].type).to.be.equal('Meccan');
    });
  });

  describe('Quran.juz(...)', () => {

    it('should return info about a juz', async () => {
      // Fetch info about juz 28
      const quran = new Quran();
      const juz = await quran.juz(28);

      // TEST: We are getting metadata for 1 juz
      expect(juz.length).to.be.equal(1);

      // TEST: 28rd juz starts from surah 58, ayat 1
      expect(juz[0].surah).to.be.equal(58);
      expect(juz[0].ayah).to.be.equal(1);
    });

    it('should return info about all juz', async () => {
      // Fetch info about all juz
      const quran = new Quran();
      const juz = await quran.juz();

      // TEST: There are 30 juz, but this query returns 31 entires
      // since it demarcates when the start and end
      expect(juz.length).to.be.equal(31);

      // TEST: Last juz starts from surah 78, ayat 1
      expect(juz[29].surah).to.be.equal(78);
      expect(juz[29].ayah).to.be.equal(1);

      // TEST: Last juz ends at the end of the quran (surah 114)
      expect(juz[30].surah).to.be.equal(115);
      expect(juz[30].ayah).to.be.equal(1);
    });

    it('should allow querying for verses in the juz', async () => {
      // Fetch info about juz 28 (i.e. where it starts)
      const quran = new Quran();
      const juz = await quran.juz(28);

      console.log(`** Juz 28: ${JSON.stringify(juz)}`);

      // Now fetch top 10 verses from verses in the juz
      const verses = await quran.select({ chapter: juz[0].surah },
        { offset: juz[0].ayah - 1, limit: 10 });

      // TEST: There are 10 verses fetched
      expect(verses.length).to.be.equal(10);

      // TEST: Arabic text of the 6th verse of the juz matches
      expect(verses[5].ar).to.startsWith('يَوْمَ يَبْعَثُهُمُ ٱللَّهُ');
    });
  });

  describe('Quran.select(...)', () => {

    it('should return a full numbered chapter in arabic', async () => {
      // Fetch Al Fatiha via select
      const quran = new Quran();
      const verses = await quran.select({ chapter: 1 });

      // TEST: Surah Fatiha has 7 ayahs
      expect(verses.length).to.be.equal(7);

      // TEST: Arabic text of the 3rd ayah of Surah Fatiha matches
      expect(verses[2].ar).to.be.equal('ٱلرَّحْمَٰنِ ٱلرَّحِيمِ');
    });

    it('should return a full numbered chapter with a translation', async () => {
      // Fetch Al Fatiha via select with a translation
      const quran = new Quran();
      const verses = await quran.select({ chapter: 1 }, {}, ['ur']);

      // TEST: Surah Fatiha has 7 ayahs
      expect(verses.length).to.be.equal(7);

      // TEST: Arabic text of the 3rd ayah of Surah Fatiha matches
      expect(verses[2].ar).to.be.equal('ٱلرَّحْمَٰنِ ٱلرَّحِيمِ');

      // TEST: Urdu text of the 3rd ayah of Surah Fatiha matches
      expect(verses[2].ur).to.be.equal('وہ عظیم اوردائمی رحمتوں والا ہے');
    });

    it('should return top ayats in arabic', async () => {
      // Fetch Al Fatiha via select with top options
      const quran = new Quran();
      const verses = await quran.select({ chapter: 1 }, { limit: 3 });

      // TEST: Surah Fatiha has 7 ayahs, of which we are fetching 3
      expect(verses.length).to.be.equal(3);

      // TEST: Arabic text of the 3rd ayah of Surah Fatiha matches
      expect(verses[2].ar).to.be.equal('ٱلرَّحْمَٰنِ ٱلرَّحِيمِ');
    });

    it('should return some select ayats in multiple languages', async () => {
      // Fetch Al Fatiha via select with top options
      const quran = new Quran();
      const verses = await quran.select({ chapter: 1, verse: [2, 4, 6] }, {}, ['en', 'ur']);

      // TEST: Surah Fatiha has 7 ayahs, of which we are fetching 3
      expect(verses.length).to.be.equal(3);

      // TEST: Urdu text of the 4th ayah of Surah Fatiha matches
      expect(verses[1].ur).to.be.equal('روزِقیامت کا مالک و مختار ہے');

      // TEST: English text of the 4th ayah of Surah Fatiha matches
      expect(verses[1].en).to.be.equal('Master of the Day of Judgment.');
    });
  });
});


test/GreetingSpec.js
/**
 * quran-promise
 *
 * Copyright © 2016 Axis, the Information Professionals. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import { expect } from 'chai';
import Greeting from '../src/Greeting';

describe('Greeting', () => {

  describe('greeting.hello()', () => {

    it('should return welcome message for a guest user', () => {
      const greeting = new Greeting();
      const message = greeting.hello();
      expect(message).to.be.equal('Welcome, Guest!');
    });

    it('should return welcome message for a named user', () => {
      const greeting = new Greeting('John');
      const message = greeting.hello();
      expect(message).to.be.equal('Welcome, John!');
    });
  });
});


src/Quran.js
/**
 * quran-promise
 *
 * Copyright © 2016 Axis, the Information Professionals. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import appRootDir from 'app-root-dir';
import db from 'sqlite';
import util from 'util';

// Defines file paths
const rootDir = appRootDir.get();
const databaseFilePath = `${rootDir}/db.sqlite`;

class Quran {
  // Helper method to get arabic verses from a chapter
  async get(chapterId, verseId = undefined) {
    // use the low level select function to query the verses
    const response = await this.select({ chapter: chapterId, verse: verseId });

    // return arabic strings only (map from array)
    let verses;
    if (!!response && response.length > 0) {
      verses = response.map(x => x.ar);
    }

    return verses;
  }

  // Helper method to get information about a chapter
  // Note: If no chapterId is specified then this method returns information
  //       about all chapters.
  async chapter(chapterId) {
    let query = 'SELECT * FROM chapters ';

    if (chapterId && chapterId > 114) {
      throw new Error(`chapterId specified is out of bounds: ${chapterId}.`);
    }

    if (!!chapterId) {
      if (!util.isNumber(chapterId)) {
        throw new Error(`Invalid chapterId: ${chapterId}`);
      }

      query += `WHERE id=${chapterId}`;
    }

    return await db.all(query);
  }

  // Helper method to get information about a juz
  // Note: If no juzId is specified then this method returns information
  //       about all juz.
  async juz(juzId) {
    let query = 'SELECT * FROM juz ';

    if (juzId && juzId > 30) {
      throw new Error(`JuzId specified is out of bounds: ${juzId}.`);
    }

    if (!!juzId) {
      if (!util.isNumber(juzId)) {
        throw new Error(`Invalid juzId: ${juzId}`);
      }

      query += `WHERE id=${juzId}`;
    }

    return await db.all(query);
  }

  // Helper method to do a free-form text search of the Quran
  async search(language, text) {
    const query = `SELECT chapter, verse, ${language} FROM ${language}` +
      `WHERE ${language} LIKE "%${text}%";`;

    return db.all(query);
  }

  // Full low level select method. Allows you to specify:
  // 1. Filters: Any array to filter on, e.g. { verse: 2, chapter: 5 }
  // 2. Options for offset and limit, e.g. { limit: 10, offset: 5 }
  // 3. Translations: An array of translations to return, e.g. ['en', 'hi']
  async select(filters, options, translations = ['ar']) {
    let query = 'SELECT * FROM ar a ';
    const params = [];

    // join translations from other tables, if any are requested
    // beyond arabic
    if (!(translations.length === 1 && translations[0] === 'ar')) {
      translations.forEach(l => {
        query += `JOIN ${l} USING(chapter, verse) `;
      });
    }

    // add filters (where clauses)
    if (!!filters) {
      Object.keys(filters).forEach(k => {
        const f = filters[k];

        if (!!f) {
          if (util.isArray(f)) {
            params.push(`a.${k} IN (${f.join(',')})`);
          } else {
            params.push(`a.${k}=${f}`);
          }
        }
      });

      query += `WHERE ${params.join(' AND ')}`;
    }

    // apply sorting, offset and limit options
    query += ' ORDER BY chapter, verse ';
    if (!!options) {
      ['limit', 'offset'].forEach(x => {
        if (options[x]) {
          query += `${x} ${options[x]} `;
        }
      });
    }

    await db.open(databaseFilePath);
    return await db.all(query);
  }
}

export default Quran;


tools/build.js
/**
 * Babel Starter Kit (https://www.kriasoft.com/babel-starter-kit)
 *
 * Copyright © 2015-2016 Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

import fs from 'fs-promise';
import del from 'del';
import rollup from 'rollup';
import babel from 'rollup-plugin-babel';
import pkg from '../package.json';

// Transpile source code into a distributable format with Babel.
async function transpile() {
  for (const format of ['es6', 'cjs', 'umd']) {
    // bundle the source.
    const bundle = await rollup.rollup({
      entry: 'src/index.js',
      external: Object.keys(pkg.dependencies),
      plugins: [babel(Object.assign(pkg.babel, {
        babelrc: false,
        exclude: 'node_modules/**',
        runtimeHelpers: true,
        presets: pkg.babel.presets.map(x => (x === 'es2015' ? 'es2015-rollup' : x)),
      }))],
    });

    // write the bundle, generating source maps.
    await bundle.write({
      dest: `dist/${format === 'cjs' ? 'index' : `index.${format}`}.js`,
      format,
      sourceMap: true,
      moduleName: format === 'umd' ? pkg.name : undefined,
    });
  }
}

// Copy package files: database, package.json, README and LICENSE.txt.
async function copyPackageFiles() {
  delete pkg.private;
  delete pkg.devDependencies;
  delete pkg.scripts;
  delete pkg.eslintConfig;
  delete pkg.babel;
  await fs.writeFile('dist/package.json', JSON.stringify(pkg, null, '  '), 'utf-8');
  await fs.writeFile('dist/LICENSE.txt', await fs.readFile('LICENSE.txt', 'utf-8'), 'utf-8');
  await fs.writeFile('dist/README.md', await fs.readFile('README.md', 'utf-8'), 'utf-8');
  await fs.createReadStream('./src/db.sqlite').pipe(await fs.createWriteStream('dist/db.sqlite'));
}

// execute as an IIFE: Immediately-Invoked Function Expression (IIFE)
(async function main() {
  try {
    // Clean up the output directory
    del(['dist/*']);

    // transpile with babel
    await transpile();

    // copy package files
    await copyPackageFiles();
  } catch (error) {
    console.error(`Error: ${error}`); // eslint-disable-line no-console
  }
}());


tools/downloadCorpus.js
/**
 * quran-promise
 *
 * Copyright © 2016 Axis, the Information Professionals. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

import request from 'request-promise';
import fs from 'fs-promise';
import del from 'del';

// defines the output directory
const corpusDir = `${__dirname}/../corpus`;

// defines data to download from tanzil
const translations = ['en.shakir', 'ur.jawadi', 'hi.hindi'];
const quranType = 'uthmani';

// helper method to download a file via GET to a url, overwriting on any conflicts
async function getFile(url, filePath) {
  console.log(`${url} >> ${filePath}`);

  // download
  const contents = await request.get(url);

  // write file
  await fs.writeFile(filePath, contents);
}

// helper method to download the arabic corpus
async function downloadArabic() {
  console.log('downloading Arabic corpus...');

  // download the arabic metadata from tanzil

  await getFile('http://tanzil.net/res/text/metadata/quran-data.js', `${corpusDir}/quran-data.js`);

  // append the module export to the downloaded metadata file
  await fs.appendFile(`${corpusDir}/quran-data.js`, 'module.exports = QuranData;');

  // download the arabic corpus from tanzil
  const url = 'http://tanzil.net/pub/download/download.php';
  const filePath = `${corpusDir}/quran-${quranType}.txt`;

  const options = {
    method: 'POST',
    uri: url,
    form: {
      quranType,
      marks: true,
      sajdah: true,
      alef: true,
      outType: 'txt-2',
      agree: true,
    },
  };

  const contents = await request.post(options);
  console.log(`${url} >> ${filePath}`);
  await fs.writeFile(filePath, contents);
}

// helper method to download the translation corpus for a language
async function downloadTranslation(language) {
  console.log(`downloading Translation corpus for: ${language}`);

  // download the translation from tanzil
  await getFile(`http://tanzil.net/trans/${language}`, `${corpusDir}/${language}.txt`);
}

// execute as an IIFE: Immediately-Invoked Function Expression (IIFE)
(async function main() {
  try {
    // delete contents of output directory
    del([`${corpusDir}/*`]);

    // Download Arabic Corpus from tanzil
    await downloadArabic();

    // Download Translations from tanzil
    for (const translation of translations) {
      await downloadTranslation(translation);
    }
  } catch (error) {
    console.error(`Error: ${error}`);
  }
}());


tools/downloadAndImportCorpus.js
/**
 * quran-promise
 *
 * Copyright © 2016 Axis, the Information Professionals. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

import appRootDir from 'app-root-dir';
import db from 'sqlite';
import del from 'del';
import fs from 'fs-promise';
import request from 'request-promise';
import QuranData from '../corpus/quran-data';

// defines data to download from tanzil
const translations = ['en.shakir', 'ur.jawadi', 'hi.hindi'];
const quranType = 'uthmani';

// defines the file paths
const rootDir = appRootDir.get();
const corpusDir = `${rootDir}/corpus`;
const arabicCorpusFilePath = `${corpusDir}/quran-${quranType}.txt`;
const databaseFilePath = `${rootDir}/src/db.sqlite`;

// helper method to download a file via GET to a url, overwriting on any conflicts
async function getFile(url, filePath) {
  console.log(`${url} >> ${filePath}`);

  // download
  const contents = await request.get(url);

  // write file
  await fs.writeFile(filePath, contents);
}

// helper method to download the arabic corpus
async function downloadArabic() {
  console.log('Downloading Arabic corpus...');

  // download the arabic metadata from tanzil
  await getFile('http://tanzil.net/res/text/metadata/quran-data.js', `${corpusDir}/quran-data.js`);

  // append the module export to the downloaded metadata file
  await fs.appendFile(`${corpusDir}/quran-data.js`, 'module.exports = QuranData;');

  // download the arabic corpus from tanzil
  const url = 'http://tanzil.net/pub/download/download.php';

  const options = {
    method: 'POST',
    uri: url,
    form: {
      quranType,
      marks: true,
      sajdah: true,
      alef: true,
      outType: 'txt-2',
      agree: true,
    },
  };

  const contents = await request.post(options);
  console.log(`${url} >> ${arabicCorpusFilePath}`);
  await fs.writeFile(arabicCorpusFilePath, contents);
}

// helper method to download the translation corpus for a language
async function downloadTranslation(language) {
  console.log(`Downloading Translation corpus for: ${language}`);

  // download the translation from tanzil
  await getFile(`http://tanzil.net/trans/${language}`, `${corpusDir}/${language}.txt`);
}

// helper method to import the arabic corpus into the database
async function importArabic() {
  console.log(`Importing Arabic corpus from: ${arabicCorpusFilePath}`);

  // read the raw corpus text file into in-memory rows
  // Each row is an array of chapter, verse and arabic text,
  // e.g. [ 1, 1, 'بِسْمِ ٱللَّهِ ٱلرَّحْمَٰنِ ٱلرَّحِيمِ' ]
  const rows = [];
  const contents = await fs.readFile(arabicCorpusFilePath);

  const lines = contents.toString().split('\n');

  for (let line of lines) {
    line = line.trim();
    if (!line || line.match(/^#/)) {
      continue;
    }

    const row = line.split('|');
    rows.push(row);
  }

  // now that we have parsed out rows, create the sqlite database and 'ar' table
  const tableName = 'ar';
  await db.open(databaseFilePath);
  await db.exec(`CREATE TABLE ${tableName} (chapter INTEGER, verse INTEGER, ${tableName} TEXT)`);

  // Insert the arabic text into the 'ar' table, 1k rows at a time
  const insertArabicStatementCommand = `INSERT INTO ${tableName} VALUES (?,?,?)`;
  let insertStatement = await db.prepare(insertArabicStatementCommand);
  for (let i = 0; i < rows.length; i++) {
    const row = rows[i];

    // skip invalid rows
    if (!row || row.length !== 3) continue;

    await insertStatement.run(row);
    if (i % 1000 === 0) {
      // after every 1k statements, execute and the re-initialize the statement
      await insertStatement.finalize();
      insertStatement = await db.prepare(insertArabicStatementCommand);
    }
  }

  // Create the chapters table, using the surahs metadata in QuranData.js
  await db.exec('CREATE TABLE chapters (id INTEGER, start INTEGER, ayas INTEGER, ord INTEGER, ' +
    'rukus INTEGER, arname TEXT, tname TEXT, enname TEXT, type TEXT);');

  // Build and run the statement to insert surah metadata into the chapters table
  const insertChaptersStatement = await db.prepare('INSERT INTO chapters ' +
    'VALUES (?,?,?,?,?,?,?,?,?)');

  for (let i = 0; i < QuranData.Sura.length; i++) {
    const surahInfo = QuranData.Sura[i];

    // Add the ID field at the beginning and run the statement
    surahInfo.unshift(i);

    // skip invalid rows
    if (!surahInfo || surahInfo.length !== 9) continue;

    // run
    await insertChaptersStatement.run(surahInfo);
  }

  // execute insert chapters
  await insertChaptersStatement.finalize();

  // Create the juz table, using the juz metadata in QuranData.js
  await db.exec('CREATE TABLE juz (id INTEGER, surah INTEGER, ayah INTEGER)');

  // Build and run the statement to insert juz metadata into the juz table
  const insertJuzStatement = await db.prepare('INSERT INTO juz VALUES (?,?,?)');
  for (let i = 0; i < QuranData.Juz.length; i++) {
    const juzInfo = QuranData.Juz[i];

    // Add the ID field at the beginning and run the statement
    juzInfo.unshift(i);

    // skip invalid rows
    if (!juzInfo || juzInfo.length !== 3) continue;

    // run
    await insertJuzStatement.run(juzInfo);
  }

  // execute insert juz
  await insertJuzStatement.finalize();

  // Create the page table, using the page metadata in QuranData.js
  await db.exec('CREATE TABLE page (page INTEGER, ayah INTEGER)');

  // Build and run the statement to insert page metadata into the page table
  const insertPageStatement = await db.prepare('INSERT INTO page values (?,?)');
  for (let i = 0; i < QuranData.Page.length; i++) {
    const pageInfo = QuranData.Page[i];

    // skip invalid rows
    if (!pageInfo || pageInfo.length !== 2) continue;

    // run
    await insertPageStatement.run(pageInfo);
  }

  // execute insert page
  await insertPageStatement.finalize();
}


// helper method to import a translation into the database
async function importTranslation(translation) {
  console.log(`Importing Translation: ${translation}`);

  // read the raw corpus text file into in-memory rows
  // Each row is an array of chapter, verse and english text,
  // e.g. [ 1, 1, In the name of Allah, the Beneficent, the Merciful ]
  const rows = [];
  const contents = await fs.readFile(`${corpusDir}/${translation}.txt`);

  const lines = contents.toString().split('\n');

  for (let line of lines) {
    line = line.trim();
    if (!line || line.match(/^#/)) {
      continue;
    }

    const row = line.split('|');
    if (row.length > 3) {
      // some translations span two columns, so we append them.
      row[2] += row.pop();
    }

    rows.push(row);
  }

  // now that we have parsed out rows, create the sqlite database and translation table
  const tableName = translation.split('.')[0];
  await db.open(databaseFilePath);
  await db.exec(`CREATE TABLE ${tableName} (chapter INTEGER, verse INTEGER, ${tableName} TEXT)`);

  // Insert the translation text into the translation table, 1k rows at a time
  const insertTranslationStatementCommand = `INSERT INTO ${tableName} VALUES (?,?,?)`;
  let insertStatement = await db.prepare(insertTranslationStatementCommand);
  for (let i = 0; i < rows.length; i++) {
    const row = rows[i];

    // skip invalid rows
    if (!row || row.length !== 3) continue;

    await insertStatement.run(row);
    if (i % 1000 === 0) {
      // after every 1k statements, execute and the re-initialize the statement
      await insertStatement.finalize();
      insertStatement = await db.prepare(insertTranslationStatementCommand);
    }
  }
}

// execute as an IIFE: Immediately-Invoked Function Expression (IIFE)
(async function main() {
  try {
    // delete contents of output directory
    del([`${corpusDir}/*`]);

    // Download Arabic corpus from tanzil
    await downloadArabic();

    // Import downloaded corpus into the database
    await importArabic();

    // Download and Import Translations from tanzil
    for (const language of translations) {
      // Download translation for language
      await downloadTranslation(language);

      // Import translation for language
      await importTranslation(language);
    }
  } catch (error) {
    console.error(`Error: ${error}`);
  }
}());


src/index.js
/**
 * quran-promise
 *
 * Copyright © 2016 Axis, the Information Professionals. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */
import db from 'sqlite';
import util from 'util';

// Defines file paths
const databaseFilePath = `${__dirname}/db.sqlite`;

class Quran {
  // Helper method to get arabic verses from a chapter
  async get(chapterId, verseId = undefined) {
    // use the low level select function to query the verses
    const response = await this.select({ chapter: chapterId, verse: verseId });

    // return arabic strings only (map from array)
    let verses;
    if (!!response && response.length > 0) {
      verses = response.map(x => x.ar);
    }

    return verses;
  }

  // Helper method to get information about chapters
  // Note: If no chapterId is specified then this method returns information
  //       about all chapters.
  async chapters(chapterId) {
    let query = 'SELECT * FROM chapters ';

    // sanitize inputs
    if (!!chapterId) {
      if (chapterId > 114) {
        throw new Error(`chapterId specified is out of bounds: ${chapterId}.`);
      }

      if (isNaN(chapterId)) {
        throw new Error(`Invalid chapterId: ${chapterId}`);
      }

      query += `WHERE id=${chapterId}`;
    }
    await db.open(databaseFilePath);
    return await db.all(query);
  }

  // Helper method to get information about a juz
  // Note: If no juzId is specified then this method returns information
  //       about all juz.
  async juz(juzId) {
    let query = 'SELECT * FROM juz ';

    // sanitize inputs
    if (!!juzId) {
      if (juzId > 30) {
        throw new Error(`juzId specified is out of bounds: ${juzId}.`);
      }

      if (isNaN(juzId)) {
        throw new Error(`Invalid juzId: ${juzId}`);
      }

      query += `WHERE id=${juzId}`;
    }

    await db.open(databaseFilePath);
    return await db.all(query);
  }

  // Helper method to do a free-form text search of the Quran
  async search(language, text) {
    const query = `SELECT chapter, verse, ${language} FROM ${language}` +
      `WHERE ${language} LIKE "%${text}%";`;

    await db.open(databaseFilePath);
    return await db.all(query);
  }

  // Full low level select method. Allows you to specify:
  // 1. Filters: Any array to filter on, e.g. { verse: 2, chapter: 5 }
  // 2. Options for offset and limit, e.g. { limit: 10, offset: 5 }
  // 3. Translations: An array of translations to return, e.g. ['en', 'hi']
  async select(filters, options, translations = ['ar']) {
    let query = 'SELECT * FROM ar a ';
    const params = [];

    // join translations from other tables, if any are requested
    // beyond arabic
    if (!(translations.length === 1 && translations[0] === 'ar')) {
      translations.forEach(l => {
        query += `JOIN ${l} USING(chapter, verse) `;
      });
    }

    // add filters (where clauses)
    if (!!filters) {
      Object.keys(filters).forEach(k => {
        const f = filters[k];

        if (!!f) {
          if (util.isArray(f)) {
            params.push(`a.${k} IN (${f.join(',')})`);
          } else {
            params.push(`a.${k}=${f}`);
          }
        }
      });

      query += `WHERE ${params.join(' AND ')}`;
    }

    // apply sorting, offset and limit options
    query += ' ORDER BY chapter, verse ';
    if (!!options) {
      ['limit', 'offset'].forEach(x => {
        if (options[x]) {
          query += `${x} ${options[x]} `;
        }
      });
    }

    await db.open(databaseFilePath);
    return await db.all(query);
  }
}

export default Quran;
